main.go

```diff
@@ -1,83 +1,37 @@
 package main
 
 import (
-	"bufio"
 	"fmt"
-	"os"
 )
 
-// Data structure = linked list as a stack [data+head]->[2]->[1]->[0]
-
-// acts as head of list & contains DB
-type HeadDB struct {
-	// access to stack of operations
-	current *Node
-	// most recent ops for same key
-	data map[string]string
-}
-
-// represents transaction with a hash of operations
-type Node struct {
-	parent *Node
-	ops    map[string]struct
-}
-
-type Read struct {
-}
-
-type Write struct {
-	value string
-}
-
-type Delete struct {
-}
-
-// might need an interface for operations (?)
-
-func (h *HeadDB) WriteTo(k,v string) *HeadDB {
-	// reference new Write Struct
-	writeOp := Write{value: v}
-	// access current ops list
-	list = &h.current.ops
-	*list[k] = writeOp
+type Transaction struct {
+	parent *Transaction
+	data   map[string]string
 }
 
-func (h *HeadDB) Start() *HeadDB {
-	newOps := make(map[string]Op)
-	// newNode where parent is current
-	newNode := &Node{parent: h.current ops: newOps }
-	// Then reset Head's current to new node
-	h.current = newNode
-	return h
+func NewTransaction(parent *Transaction) *Transaction {
+	m := make(map[string]string)
+	return &Transaction{data: m}
 }
 
-func (h *HeadDB) Abort() *HeadDB {
-	// TODO: move the head pointer and delete current Node
-	return h
+func (n *Transaction) Write(k, v string) {
+	n.data[k] = v
 }
 
-func (h *HeadDB) Commit() *HeadDB {
-	// TODO: write current commands to datastore
-	return h
+func (t *Transaction) Read(k string) string {
+	if v, ok := t.data[k]; ok {
+		return v
+	}
+	return "Key not found: " + k
```

Comment:
the idiomatic Go way to do this is to use the `error` type. the Read method will return (string, error). if you find the value, return it and a nil error. If you don't find it, return an empty string and an error with the message you have. The error return would look something like:

`return "", fmt.Errorf("Key not found: %s", k)`